com.minres.coredsl.tests/src/com/minres/coredsl/tests/CoreDslAttributeValidationTest.xtend

```diff
@@ -0,0 +1,202 @@
+import org.eclipse.xtext.testing.util.ParseHelper
+import org.eclipse.xtext.testing.validation.ValidationTestHelper
+import org.junit.jupiter.api.Test
+import org.junit.jupiter.api.^extension.ExtendWith
+
+import static org.junit.jupiter.api.Assertions.assertEquals
+import static org.junit.jupiter.api.Assertions.assertNotNull
+import static org.junit.jupiter.api.Assertions.assertTrue
+
+import com.minres.coredsl.validation.KnownAttributes
+import com.minres.coredsl.validation.CoreDslValidator
+
+@ExtendWith(InjectionExtension)
+@InjectWith(CoreDslInjectorProvider)
+class CoreDslAttributeValidationTest {
+
+	@Inject extension ParseHelper<DescriptionContent> parseHelper
+
+	@Inject ValidationTestHelper validator
+
+	@Test
+	def void validAttributePlacement() {
+		val content = '''
+			InstructionSet AttributeValidationTestInstructionSet {
+			    architectural_state {
+			        unsigned XLEN;
+			        unsigned<XLEN> reg[32] [[is_main_reg]];
+			        unsigned<XLEN> mem[1024] [[is_main_mem]];
+			        unsigned<XLEN> pc [[is_pc]];
+			        unsigned<XLEN> lock [[is_interlock_for(pc)]];
+			    }
+			    functions {
+			    	void fun() [[do_not_synthesize]] {}
+			    }
+			    instructions [[enable=true]] [[hls]] {
+			    	test [[hls]] [[no_cont]] [[cond]] [[flush]] {
+			    		encoding: 0;
+			    		args_disass: "";
+			    		behavior: {}
+			    	}
+			    }
+			}
+			Core AttributeValidationTestCore provides AttributeValidationTestInstructionSet {
+			    architectural_state {
+			        XLEN=32;
+			    }
+			}
+		'''.parse();
+
+		val issues = validator.validate(content);
+		for (issue : issues) {
+			println(issue);
+		}
+
+		assertTrue(issues.isEmpty());
+	}
+
+	@Test
+	def void invalidAttributePlacement() {
+		val attribString = KnownAttributes.all.map [
+			"[[" + it.name + (it.paramCount > 0 ? "(" + ", 0".repeat(it.paramCount).substring(2) + ")" : ""
+    	) + "]]"
+		].join(" ");
+
+		val content = '''
+			InstructionSet AttributeValidationTestInstructionSet {
+			    architectural_state {
+			        unsigned XLEN;
+			        unsigned<XLEN> field «attribString»;
+			    }
+			    functions {
+			    	void fun() «attribString» {}
+			    }
+			    instructions «attribString» {
+			    	test «attribString» {
+			    		encoding: 0;
+			    		args_disass: "";
+			    		behavior: {}
+			    	}
+			    }
+			}
+			Core AttributeValidationTestCore provides AttributeValidationTestInstructionSet {
+			    architectural_state {
+			        XLEN=32;
+			    }
+			}
+		'''.parse();
+
+		val issues = validator.validate(content);
+		for (issue : issues) {
+			println(issue);
+			assertEquals(issue.code, CoreDslValidator.ILLEGAL_ATTRIBUTE);
+		}
+
+		// Every attribute appears in every possible location, so we expect
+		// <location count> * <attribute count> - <valid placement count> issues.
+		val expectedIssueCount = KnownAttributes.all.map[4 - it.allowedUsage.size].reduce[a, b|a + b];
+		assertEquals(issues.size, expectedIssueCount);
```

Comment:
I think I know where you coming from here (these issues are encoded in a weird, deep data structure, right?), but still this check seems susceptible to unrelated errors messing/hiding the one we are looking for. Could this be a simple string-search on the issues' `toString()`? (Again, it would be ok if you'd pick one or two attributes to check in the unit test; this doesn't have to be exhaustive.)

Reply:
The loop immediately above checks that all issues are of type `CoreDslValidator.ILLEGAL_ATTRIBUTE`, so any unrelated errors would cause the test to fail.

Reply:
Yup, overlooked that.